Improve resiliency of entities when destroyed

wayneparrott opened the pull request:
Issue #896 identifies an race condition where multiple subscriptions listen to a common topic and one subscription, from it's callback destroys another subscription (an entity). This revealed an issue where i/o activity can exists on an entity's _handle after the entity has been destroyed. The result is an invalid entity pointer or worse a SEGFAULT. To address this I've added destroy state semantics to the Entity class. 

full disclosure:
One quirky behavior I've attempted to resolve is that setting `Entity._handle = null`, results in the _handle being quickly GC'ed. This causes the SEGFAULT when the handle is still referenced by ShadowNode.Execute(). I believe after calling _handle.release(), the handle must not be GC'ed until after the current Node.execute() completes. I was not sure how to pull this off in C with Nan API so I implemented a hack by using a cache to hold `_handle`'s after their `release()`. The cache will flush when its size exceeds a const (100 atm). I'm definitely open to revising this hack with the suggestion of others.

entity.js
- added _destroy(), isDestroyed()
- added cache to hold obsolete _handles to avoid premature GC

entity.d.ts
- added isDestroyed()

node.js
- added guard statements to execute() to filter out destroyed entities

action/client.js
- added _destory() and isDestroyed()
- refactored destroy() flow

action/server.js
- added _destory() and isDestroyed()
- refactored destroy() flows

main.ts
- added tests for isDestroyed() to entity subclasses

test-destruction.js
- added entity.destory() test cases

Fix #896

wayneparrott committed:
improve entity destroy() resiliency #896

entity.js
- added _destroy(), isDestroyed()
- added cache to hold obsolete _handles to avoid premature GC

entity.d.ts
- added isDestroyed()

node.js
- added guard statements to execute() to filter out destroyed entities

action/client.js
- added _destory() and isDestroyed()
- refactored destroy() flow

action/server.js
- added _destory() and isDestroyed()
- refactored destroy() flows

main.ts
- added tests for isDestroyed() to entity subclasses

test-destruction.js
- added entity.destory() test cases

minggangw commented on lib/entity.js:
```diff
@@ -14,6 +14,19 @@
+const MAX_OBSOLETE_HANDLES_SIZE = 100;
+const OBSOLETE_HANDLES = new Set();
+function registerObsoleteHandle(handle) {
+  OBSOLETE_HANDLES.add(handle);
```
I think we can check if the handle has been deleted when calling `execute`, and remove if the handle has been released.

https://github.com/RobotWebTools/rclnodejs/blob/0ea952f29e6251e229cf132790d61946ca4fca7f/lib/node.js#L150-L155

like
```js
execute(handles) {
  handles.forEach(handle => {
     if (OBSOLETE_HANDLES.indexOf(handle) !== -1) {
       // Remove from OBSOLETE_HANDLES.
      }
  });
...
```
I haven't tested, just in my mind:) thus we don't have to keep these handles for a long time.

wayneparrott committed:
improve destroyed entity handling to avoid segfault

entity.js
- added static gcHandles()

node.js
- revised execute() to use for-of loops for each entity type. This
simplifies the looping of each group of entities through the use of
loop interruption (e.g., use continue statement) to detect a
destroyed entity before processing. Additionally it the for-of loop
reduces the number of isDestroyed() calls Lastly, added a call to
Entity.gcHandles() prior to exiting the method.

wayneparrott commented:
@minggangw my latest commit [9af468f](https://github.com/RobotWebTools/rclnodejs/commit/9af468f7bdc3ce4c7d0819c09d999e2162589b90) to address your suggestion may have overwritten [f3840e2](https://github.com/RobotWebTools/rclnodejs/commit/f3840e292e438bf21e3e84d187b18f64e2980caa)

minggangw commented on lib/node.js:
```diff
@@ -371,7 +382,11 @@ class Node extends rclnodejs.ShadowNode {
+
+    // At this point it is safe to clear the cache of any
+    // destroyed entity references
+    Entity._gcHandles();
```
this looks pretty good :)

minggangw commented:
> @minggangw my latest commit [9af468f](https://github.com/RobotWebTools/rclnodejs/commit/9af468f7bdc3ce4c7d0819c09d999e2162589b90) to address your suggestion may have overwritten [f3840e2](https://github.com/RobotWebTools/rclnodejs/commit/f3840e292e438bf21e3e84d187b18f64e2980caa)

Sorry for missing this comment, the change looks good to me, thanks!

minggangw reviewed (approved):
LGTM! Thanks for fixing the crashing issue

minggangw commented:
Hi @wayneparrott I have approved this PR, please go ahead to merge it after resolving the conflict, thanks!

wayneparrott committed:
Merge branch 'develop' into destroyed_entity_896
